File: project/backend/src/main/java/dev/lueem/clients/OpenAiClient.java
package dev.lueem.clients;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.util.logging.Level;
import java.util.logging.Logger;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.json.Json;
import jakarta.json.JsonArray;
import jakarta.json.JsonObject;
import jakarta.json.JsonReader;
import java.io.StringReader;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import dev.lueem.payload.PayloadBuilder;

@ApplicationScoped
public class OpenAiClient {

    @ConfigProperty(name = "OPENAI_API_KEY")
    private String openaiKey;

    private static final Logger LOGGER = Logger.getLogger(OpenAiClient.class.getName());
    private final HttpClient httpClient = HttpClient.newHttpClient();

    @Inject
    private PayloadBuilder payloadBuilder;

    private JsonArray processResponse(String responseBody) {
        try (JsonReader jsonReader = Json.createReader(new StringReader(responseBody))) {
            JsonObject jsonResponse = jsonReader.readObject();
            JsonArray choices = jsonResponse.getJsonArray("choices");
            if (choices == null || choices.isEmpty()) {
                throw new RuntimeException("No choices found in the response.");
            }
            JsonObject message = choices.getJsonObject(0).getJsonObject("message");
            if (message == null) {
                throw new RuntimeException("No message found in the first choice.");
            }
            String contentString = message.getString("content");

            LOGGER.fine("Content received from OpenAI: " + contentString);

            try (JsonReader contentReader = Json.createReader(new StringReader(contentString))) {
                JsonObject contentJson = contentReader.readObject();
                return contentJson.getJsonArray("ArticleList");
            }
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Error processing response", e);
            throw new RuntimeException("Error processing response", e);
        }
    }

    public JsonArray askQuestion(String question) {
        try {
            String payload = payloadBuilder.constructPayload(question);

            if (openaiKey == null || openaiKey.trim().isEmpty()) {
                throw new IllegalArgumentException("API KEY is not set");
            }

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("https://api.openai.com/v1/chat/completions"))
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + openaiKey)
                    .POST(HttpRequest.BodyPublishers.ofString(payload))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            return processResponse(response.body());
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Error in askQuestion", e);
            return Json.createArrayBuilder().build(); // Return an empty JsonArray on error, to complete the flow
        }
    }
}

File: project/backend/src/main/kotlin/dev/lueem/client/OpenAiClient.kt
package dev.lueem.client

import dev.lueem.model.openai.OpenAiRequest
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.json.Json
import jakarta.json.JsonArray
import jakarta.json.bind.JsonbBuilder
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.logging.Level
import java.util.logging.Logger
import org.eclipse.microprofile.config.inject.ConfigProperty

/**
 * Client for interacting with the OpenAI API to extract information from text.
 *
 * Responsibilities:
 * - Build the chat completion request
 * - Send the request
 * - Parse the structured JSON reply
 */
@ApplicationScoped
class OpenAiClient {

    @Inject
    constructor(
        requestFactory: OpenAiChatRequestFactory,
        responseParser: OpenAiArticleListParser
    ) {
        this.requestFactory = requestFactory
        this.responseParser = responseParser
    }

    private val requestFactory: OpenAiChatRequestFactory
    private val responseParser: OpenAiArticleListParser

    @ConfigProperty(name = "OPENAI_API_KEY")
    var openaiKey: String? = null

    private val httpClient = HttpClient.newHttpClient()
    private val jsonb = JsonbBuilder.create()

    companion object {
        private val LOGGER = Logger.getLogger(OpenAiClient::class.java.name)

        private const val OPENAI_CHAT_URL = "https://api.openai.com/v1/chat/completions"
    }

    /**
     * Sends a question to the OpenAI API and retrieves the answer as a JSON array.
     *
     * @param question The question or prompt to send to OpenAI.
     * @return A JsonArray containing the extracted data, or an empty array if an error occurs.
     */
    fun extractArticlesFromText(question: String): JsonArray {
        return try {
            if (openaiKey.isNullOrBlank()) {
                throw IllegalArgumentException("API KEY is not set")
            }

            val requestBody = createRequestBody(question)
            val jsonPayload = jsonb.toJson(requestBody)

            val request =
                    HttpRequest.newBuilder()
                            .uri(URI.create(OPENAI_CHAT_URL))
                            .header("Content-Type", "application/json")
                            .header("Authorization", "Bearer " + openaiKey!!)
                            .POST(HttpRequest.BodyPublishers.ofString(jsonPayload))
                            .build()

            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            val status = response.statusCode()
            if (status < 200 || status >= 300) {
                LOGGER.severe("OpenAI HTTP $status.")
                return Json.createArrayBuilder().build()
            }

            responseParser.parseArticleList(response.body(), jsonb)
        } catch (e: Exception) {
            LOGGER.log(Level.SEVERE, "Error in extractArticlesFromText", e)
            Json.createArrayBuilder().build()
        }
    }

    /**
     * Creates the request body for the OpenAI API call.
     *
     * @param question The user's question to be included in the request.
     * @return An OpenAiRequest object containing the model, messages, and temperature.
     */
    private fun createRequestBody(question: String): OpenAiRequest {
        return requestFactory.buildChatCompletionRequest(question)
    }
}
